src/snowflake/cli/_plugins/workspace/entity_commands.py

```diff
@@ -0,0 +1,204 @@
+    ):
+        super().__init__(name=name, help=help_text, is_hidden=lambda: True)
+        self.target_id = target_id
+        self._tree_path = tree_path
+        self._command_map = {}
+        self._subtree_map = {}
+
+    def command(self, name: str, *args, **kwargs):
+        """Assume the first arg is the command name, unlike superclass"""
+
+        def decorator(command):
+            cmd_data = SnowTyperCommandData(command, args=[name, *args], kwargs=kwargs)
+            self.commands_to_register.append(cmd_data)
+            self._command_map[name] = cmd_data
+            return command
+
+        return decorator
+
+    def create_instance(self) -> SnowTyper:
+        """Provides a default help value generated based on sub-groups and commands."""
+        if not self.help:
+            subcommands = sorted(
+                [
+                    *[f"`{x}`" for x in self._subtree_map.keys()],
+                    *self._command_map.keys(),
+                ]
+            )
+            self.help = r"\+ " + ", ".join(subcommands)
+
+        return super().create_instance()
+
+    def new_subtree(self, atom: str) -> "EntityCommandGroup":
+        if atom in self._subtree_map:
+            logger.error("Duplicate subtree attempted to be created: %s", atom)
+        else:
+            subtree = EntityCommandGroup(
+                atom, target_id=self.target_id, tree_path=[*self._tree_path, atom]
+            )
+            self._subtree_map[atom] = subtree
+            self.add_typer(subtree)
+
+        return self._subtree_map[atom]
+
+    def _get_subtree(self, group_path: list[str]) -> "EntityCommandGroup":
+        """
+        Gets a group subtree factory for a sub-tree of this command group.
+        Creates groups on-the-fly (i.e. mkdir -p semantics).
+        """
+        subtree = self
+        for atom in group_path:
+            if atom in self._subtree_map:
+                subtree = self._subtree_map[atom]
+            else:
+                subtree = subtree.new_subtree(atom)
+        return subtree
+
+    def register_command_leaf(
+        self, name: str, entity_type: Entity, action: EntityAction
+    ):
+        """Registers the provided action at the given name"""
+        fn = entity_type.get_action_callable(action)
+        sig = inspect.signature(fn)
+        known_params = list(sig.parameters.keys())
+
+        @with_project_definition()
+        def _action_executor(**options) -> CommandResult:
+            cli_context = get_cli_context()
+            ws = WorkspaceManager(
+                project_definition=cli_context.project_definition,
+                project_root=cli_context.project_root,
+            )
+
+            # filter out common CLI options our action doesn't take
+            pass_through_options = {
+                k: v for (k, v) in options.items() if k in known_params
+            }
+            ws.perform_action(self.target_id, action, **pass_through_options)
```

Comment:
how do we handle return values? for example, we currently have some command that wrap the return value of the entity action in `MessageResult`, others use `QueryResult`, etc